watertrading/code/BuildModel.py

```diff
@@ -0,0 +1,1361 @@
+
+
+##################################################
+# CREATE MODEL FUNCTION
+def create_model(restricted_set, df_producer, df_consumer, df_midstream, df_distance, df_time, default={}):
+    model = ConcreteModel()
+
+    # import config dictionary
+    model.config = CONFIG(default)
+    model.type = "water_clearing"
+
+    # enable use of duals
+    model.dual = Suffix(direction=Suffix.IMPORT)
+
+    # add data frames to model
+    model.df_producer = df_producer
+    model.df_consumer = df_consumer
+    model.df_midstream = df_midstream
+    model.df_distance = df_distance
+    model.df_time = df_time
+    
+    pyunits.load_definitions_from_strings(["USD = [currency]"])
+
+    model.model_units = {
+        "volume": pyunits.oil_bbl,
+        "distance": pyunits.mile,
+        #"diameter": pyunits.inch,
+        #"concentration": pyunits.kg / pyunits.liter,
+        "time": pyunits.day,
+        "volume_time": pyunits.oil_bbl / pyunits.day,
+        "currency":pyunits.USD,
+        "currency_volume": pyunits.USD / pyunits.oil_bbl,
+        "currency_volume-distance": pyunits.USD /(pyunits.oil_bbl*pyunits.mile),
+    }
+
+    
+    # DETERMINE DATE RANGE - assume need to optimize from the earliest date to the latest date in data
+    first_date = min(df_producer["Start Date"].append(df_consumer["Start Date"]).append(df_midstream["Start Date"]))
+    last_date = max(df_producer['End Date'].append(df_consumer['End Date']).append(df_midstream["End Date"]))
+
+    # map dates to a dictionary with index
+    time_list = pd.date_range(first_date, periods=(last_date-first_date).days+1).tolist()
+    s_t = ["T_"+str(i+1) for i in range(len(time_list))] # set of time index strings
+    model.d_t = dict(zip(time_list, s_t)) # get time index from date_time value
+    model.d_T = dict(zip(s_t, time_list)) # get date_time value from time index
+    model.d_T_ord = dict(zip(s_t, range(len(time_list)))) # get ordinate from time index
+    
+    # ------------------- SETS & PARAMETERS ----------------------- #
+    # SETS
+    model.s_T = Set(
+        initialize=s_t, #NOTE: this sets up the time index based on index strings 
+        ordered=True,
+        doc="Time Periods"
+    )
+
+    model.s_PI = Set(
+        initialize=model.df_producer['Index'], 
+        doc="Producer entry index"
+    )
+    
+    model.s_CI = Set(
+        initialize=model.df_consumer['Index'], 
+        doc="Consumer entry index"
+    )
+
+    model.s_MI = Set(
+        initialize=model.df_midstream["Index"],
+        doc="Midstream entry index"
+    )
+    
+    model.s_PP = Set(
+        initialize=model.df_producer['Wellpad'], 
+        doc="Producer wellpad"
+    )
+    
+    model.s_CP = Set(
+        initialize=model.df_consumer['Wellpad'], 
+        doc="Consumer wellpad"
+    )
+
+    ### PARAMETERS
+    Producer_Wellpad = dict(zip(model.df_producer.Index, model.df_producer.Wellpad))
```

Comment:
When we initialize the set s_PI by df_producer['Index'], do we have duplicate values in them for the producers? This question arises as there is only 1 scenario where s_PI initialization and Producer_Wellpad initialization would work well. Only when the data has unique index for the producer with no duplicates and each producer has 1 well pad. If that is not the case we might be increasing the size of our set.